README.txt
Dependencies

mailtoplone.base:

    - http://pypi.python.org/pypi/mailtoplone.base

Scrawl:

    - http://pypi.python.org/pypi/Products.Scrawl

Example buildout

buildout:

    - http://svn.plone.org/svn/collective/mailtoplone/buildout-scrawl


docs/HISTORY.txt
Changelog
=========

trunk (0.3dev)
--------------

- changed category to **sent by email** [hplocher]

0.2 (2009-05-10)
----------------

- fix issue querying multidapter with None,
  now correctly use context.REQUEST
  [ramonski]

0.1 (2009-05-29)
----------------

- call add_category [hplocher]
- get all attachments from mail, and call add_file
  for each [hplocher]
  call add_file for each, before calling create
- decode subject header, used as title [hplocher]
- refactor adapter, use IBlogEntryFactory [hplocher]
- add adapter to create a Blog Entry [hplocher]
  (title, body text, only atm)



mailtoplone/scrawl/tests/base.py
"""Test setup for integration and functional tests.

When we import PloneTestCase and then call setupPloneSite(), all of
Plone's products are loaded, and a Plone site will be created. This
happens at module level, which makes it faster to run each test, but
slows down test runner startup.
"""

from Products.Five import zcml
from Products.Five import fiveconfigure

from Testing import ZopeTestCase as ztc

from Products.PloneTestCase import PloneTestCase as ptc
from Products.PloneTestCase.layer import onsetup

import Products.Scrawl
import mailtoplone.base
import mailtoplone.contentrules
import mailtoplone.scrawl

@onsetup
def setup_product():
    """Set up the package and its dependencies.

    The @onsetup decorator causes the execution of this body to be
    deferred until the setup of the Plone site testing layer. We could
    have created our own layer, but this is the easiest way for Plone
    integration tests.
    """

    # Load the ZCML configuration for the example.tests package.
    # This can of course use <include /> to include other packages.

    fiveconfigure.debug_mode = True
    zcml.load_config('configure.zcml', mailtoplone.base)
    zcml.load_config('configure.zcml', mailtoplone.contentrules)
    zcml.load_config('configure.zcml', mailtoplone.scrawl)
    fiveconfigure.debug_mode = False

    ztc.installProduct('Scrawl')
    ztc.installPackage('mailtoplone.base')
    ztc.installPackage('mailtoplone.contentrules')

# The order here is important: We first call the (deferred) function
# which installs the products we need for this product. Then, we let
# PloneTestCase set up this product on installation.

setup_product()
ptc.setupPloneSite(products=['Scrawl', 'mailtoplone.base', 'mailtoplone.contentrules'])

class TestCase(ptc.PloneTestCase):
    """We use this base class for all the tests in this package. If
    necessary, we can put common utility or setup code in here. This
    applies to unit test cases.
    """

class FunctionalTestCase(ptc.FunctionalTestCase):
    """We use this class for functional integration tests that use
    doctest syntax. Again, we can put basic common utility or setup
    code in here.
    """

    def afterSetUp(self):
        roles = ('Member', 'Contributor')
        self.portal.portal_membership.addMember('contributor',
                                                'secret',
                                                roles, [])


mailtoplone/scrawl/tests/test_doctest.py
import unittest
import doctest

from zope.testing import doctestunit
from zope.component import testing, eventtesting

from Testing import ZopeTestCase as ztc

from mailtoplone.scrawl.tests import base

def test_suite():
    return unittest.TestSuite([

        # Demonstrate the main content types
        ztc.ZopeDocFileSuite(
            'README.txt', package='mailtoplone.scrawl',
            test_class=base.FunctionalTestCase,
            optionflags=doctest.REPORT_ONLY_FIRST_FAILURE |
                doctest.NORMALIZE_WHITESPACE | doctest.ELLIPSIS),

        ])

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


mailtoplone/scrawl/adapter.py
# -*- coding: utf-8 -*-
#
# File: adapter.py
#
# Copyright (c) InQuant GmbH
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__    = """Hans-Peter Locher<hans-peter.locher@inquant.de>"""
__docformat__ = 'plaintext'

import email
import logging

from zope import interface
from zope import component

from mailtoplone.base.interfaces import IMailDropBox
from mailtoplone.base.interfaces import IBodyFactory

from mailtoplone.base.browser.emailview import EmailView

from mailtoplone.scrawl.interfaces import IBlogEntryFactory
from mailtoplone.scrawl.blogentryfactory import BlogEntryFactory
from mailtoplone.scrawl.utilities import attachments
from mailtoplone.scrawl.config import CATEGORY

info = logging.getLogger("mailtoplone.scrawl").info


class ScrawlMailDropBox(object):
    """ adapts IScrawlMailDropBoxmarker to a IMailDropBox """

    interface.implements(IMailDropBox)

    def __init__(self, context):
        self.context = context

    def drop(self, mail):
        """ drop a mail into this mail box. The mail is
        a string with the complete email content """
        body_factory = component.queryUtility(IBodyFactory)
        email_view = EmailView(self.context, None)
        entry_factory = component.queryMultiAdapter((
                self.context, self.context.REQUEST), IBlogEntryFactory,\
                default=BlogEntryFactory(self.context, None))

        # get the body and matching content_type, charset
        body, content_type, charset = body_factory(mail)

        mailobj = email.message_from_string(mail)

        # Subject
        for key in "Subject subject Betreff x-blog-title".split():
            subject = mailobj.get(key)
            if subject is not None:
                break

        #decode subject
        subject = email_view.decodeheader(subject)


        for attachment in attachments(mail):
            entry_factory.add_file(
                    attachment['data'],
                    attachment['file_name'],
                    attachment['mime_type']
                    )
        entry_factory.add_category(CATEGORY)
        entry_factory.create(subject, body)



# vim: set ft=python ts=4 sw=4 expandtab :


mailtoplone/scrawl/utilities.py
# -*- coding: utf-8 -*-
#
# File: utilities.py
#
# Copyright (c) InQuant GmbH
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__ = """Hans-Peter Locher <hans-peter.locher@inquant.de>"""
__docformat__ = 'plaintext'


import logging

import email

LOGGER="wcms.scrawl"


def info(msg):
    logging.getLogger(LOGGER).info(msg)


def attachments(mail):
    """return a generator which yields dicts like
        { "data": "some text, may also be binary for other files",
          "file_name": "muha.txt"
          "mime_type": "text/plain",
          }
    """
    # we'll make an attachment for every part of the mail which has a
    # filename parameter
    m = email.message_from_string(mail)
    parts = [item for item in m.walk() if item.get_filename() != None]
    for elem in parts:
        charset = elem.get_content_charset() or "ISO-8859-1"
        data = elem.get_payload(decode=1)
        mt= elem.get_content_type().decode(charset).encode("ascii", "ignore")
        fn= elem.get_filename().decode(charset).encode("ascii", "ignore")
        yield dict(
                data=data,
                mime_type=mt,
                file_name=fn,
                )

# vim: set ft=python ts=4 sw=4 expandtab :


mailtoplone/scrawl/interfaces.py
# -*- coding: utf-8 -*-
#
# File: interfaces.py
#
# Copyright (c) InQuant GmbH
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__ = """Hans-Peter Locher <hans-peter.locher@inquant.de>"""
__docformat__ = 'plaintext'


from zope.interface import Interface
from mailtoplone.base.interfaces import IMailDropBoxMarker

class IScrawlMailDropBoxMarker(IMailDropBoxMarker):
    """ marker interface for Scrawl mail drop boxes """

class IBlogEntryFactory(Interface):
    """ create a Blog Entry

    To create a blog entry use::

        >>> factory = queryMultiAdapter((context, request), IBlogEntryFactory, default=Default(context, request))
        >>> factory.add_category("email")
        >>> factory.add_category("foo")
        >>> factory.add_file("filedata", "text/plain", "sometext.txt")
        >>> factory.add_file(imgdata, "image/jpeg", "foo.jpg")
        >>> entry = factory.create("Some silly subject", "hey there, this is my text")

    """
    def create(title, text):
        """ create the Blog Entry on the context and return it"""

    def add_file(data, name, mime_type):
        """ add a file to the Blog Entry """

    def add_category(category):
        """ add a category """

# vim: set ft=python ts=4 sw=4 expandtab :


mailtoplone/scrawl/blogentryfactory.py
# -*- coding: utf-8 -*-
#
# File: blogentryfactory.py
#
# Copyright (c) InQuant GmbH
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__ = """Hans-Peter Locher <hans-peter.locher@inquant.de>"""
__docformat__ = 'plaintext'


import logging

from Acquisition import aq_base

from zope import interface
from zope import component
from zope.event import notify

from zope.app.container.interfaces import INameChooser
from plone.i18n.normalizer.interfaces import IIDNormalizer

from mailtoplone.base.events import MailDroppedEvent

from mailtoplone.scrawl.interfaces import IBlogEntryFactory

LOGGER="wcms.scrawl"


def info(msg):
    logging.getLogger(LOGGER).info(msg)


class BlogEntryFactory(object):
    """ create a Blog Entry

    To create a blog entry use::

        >>> factory = queryMultiAdapter((context, request), IBlogEntryFactory, default=Default(context, request))
        >>> factory.add_category("email")
        >>> factory.add_category("foo")
        >>> factory.add_file("filedata", "text/plain", "sometext.txt")
        >>> factory.add_file(imgdata, "image/jpeg", "foo.jpg")
        >>> entry = factory.create("Some silly subject", "hey there, this is my text")

    """
    interface.implements(IBlogEntryFactory)

    def __init__(self, context, request):
        self.context = context
        self.request = request

    def create(self, title, text):
        """ create the Blog Entry on the context and return it"""

        normalizer = component.getUtility(IIDNormalizer)
        chooser = INameChooser(self.context)
        id = chooser.chooseName(normalizer.normalize(title), aq_base(self.context))

        self.context.invokeFactory(
                'Blog Entry',
                id=id,
                title=title,
                text=text)

        blog_entry = getattr(self.context, id)
        blog_entry.processForm()
        notify(MailDroppedEvent(blog_entry, self.context))
        return blog_entry

    def add_file(self, data, name, mime_type):
        """ add a file to the Blog Entry """
        pass

    def add_category(self, category):
        """ add a category """
        pass



# vim: set ft=python ts=4 sw=4 expandtab :
